README.md

GridHeist
===========

A jQuery plugin to hijack an image gallery, making it great.

Check out [the dynamic test page](http://cav.is/gridheist/test/dynamic.html "Dynamic Test").  Or the [image preloading test page](http://cav.is/gridheist/test/preloader.html "Image Preloading").

![Alt text](http://cav.is/img/gridheist-dynamic-example.png "GridHeist Test")

Requirements
------------

### jQuery

Requires [jQuery](http://jquery.com/ "jQuery") version **>= 1.7.0**


Compiling
---------

If you so choose to compile/test/develop yourself, you'll need to install the `grunt-cli` using the Node Package Manager...

    npm install -g grunt-cli

Then install the dependencies from `package.json` (if you haven't already)...

    npm install <module-name-here>

Then you can run grunt tasks!

    grunt build
    grunt watch
    grunt dev


Getting Started
---------------

Include `build/gridheist.js` and `build/gridheist.css`, or some version of them.  (See example/index.html for an example).

### In your `<head>` section (next to other CSS):

    <link href="path/to/gridheist.css" rel="stylesheet">

### At the bottom of the `<body>` section (next to other JS)

    <script src="path/to/jquery-1.10.2.min.js"></script>
    <script src="path/to/gridheist.js"></script>


Usage
------------

GridHeist is a jQuery plugin, so just call the function on a selector to create a gallery.

    $(document).ready(function() {
        var options = {};
        $('#gallery').gridHeist(options);
    });

Galleries should have semi-standardized markup.  You should apply whatever layout-css you want to the outer container (`#gallery` in this example).  The following is pretty standard:

    <div id="gallery" style="margin:0 100px; width:auto;">
      <a href="img/something1_big.png">
        <img src="img/something1_thumb.png"/>
      </a>
      <a href="img/something2_big.png">
        <img src="img/something2_thumb.png"/>
      </a>
    </div>

Be aware that, for performance reasons, gridheist will wrap all the thumbnails `a` in a container `<div class="gridheist-wrap"></div>`.  So make sure your styling is scoped correctly!

If you want things to load snappier, I'd recommend pre-populating the `data-width` and `data-height` of the thumbnail images, so we don't have to wait for the `<img>` to load before we can get the native dimensions.  Note that this is the height/width of the THUMBNAIL, not of the full-sized image.

    <div id="gallery">
      <a href="img/something1_big.png">
        <img src="img/something1_thumb.png" data-width="300" data-height="200"/>
      </a>
    </div>


Options
------------

These options can be passed in when you create the gridheist.  Numbers are probably something in `px`, unless otherwise noted.

`thumbSelector` - A jQuery selector to get the thumbnail objects within the element - __(default: `> *`)__

`thumbBorder` - Border width between thumbnails - __(default: `10`)__

`thumbMinHeight` - Minimum height of a row of thumbnails - __(default: `200`)__

`thumbMaxHeight` - Maximum height of a row of thumbnails - __(default: `null`)__

`expander` - Enable the image expander - __(default: true)__

`expandRatio` - The percentage of viewport height for the expander-row - __(default: `0.60`)__

`expandMaxHeight` - Maximum height of the expander-row - __(default: `500`)__

`expandMinHeight` - Minimum height of the expander-row - __(default: `100`)__

`expandSideRender` - Function to return markup for the expander sidebar; passed the jQuery `$thumb` object - __(default: `false`)__

`expandSideWidth` - Width of the expander sidebar (if expandSideRender is passed) - __(default: `200`)__

`scroller` - Selector for the element within which the gallery will be scrolling - __(default: `window`)__

`preload` - Agressively preload full-size images when rows are expanded - __(default: `false`)__

`preloadBefore` - The number of images before the expanded row to preload - __(default: 2)__

`preloadAfter` - The number of images after the expanded row to preload - __(default: 6)__


Methods
------------

`update()` - Force a re-layout of the gallery

`update(options)` - Alter any config options, and force a re-layout of the gallery.  Options must be an object containing only the option keys you want to change.


Changes
-----------------------

Times they are a-changin'.  Check out the [changelog](CHANGELOG.md).


Issues and Contributing
-----------------------

Report any bugs or feature-requests via the issue tracker.  Or send me a fax.  And if you'd like to contribute, send me a note!  Thanks.


License
------------

GridHeist is free software, and may be redistributed under the MIT-LICENSE.

Thanks for listening!


CHANGELOG.md

GridHeist Versions
====================

0.0.1
------------

* Basic row layout functionality

0.0.2
------------

* Handle different thumb heights
* Minimum thumb height option
* Key listeners (left-right-escape)
* Scrollbar changer (when keying through)
* Better event cleanup
* Image preloading option
* Grid images can't have a max-width set (bootstrap tries to do this)

0.0.3
------------
* Bugfixes
* Add thumbMaxHeight option to scale down image heights
* Add left/right/close buttons to expander
* Unselectable buttons

0.0.4
------------
* Add `expander` option to hide the expander altogether
* Fix bug with browsers initially reporting an incorrect `$el.width()`, messing up the first layout
* Fix image-jumping when browser window gets smaller (adds an inner `.gridheist-wrap` container to the `$el`)
* Various optimizations, so complex and awesome that you totally wouldn't appreciate them

0.0.5
------------
* New preloading strategy and options for image expander
* Test for preloading

0.0.6
------------
* Dynamic expander height!
* Customizable scrolling container (doesn't always have to be `window`)
* Scroll to the correct place when opening expander (so the entire expander is visible)
* Vertically and Horizontally center images in the expander, in case they don't entirely fit


test/dynamic.js
//
// gridheist config
//
var defaults = {
  thumbBorder:      10,
  thumbMinHeight:   100,
  thumbMaxHeight:   200,
  expander:         false,
  expandHeight:     300,
  expandSideWidth:  200,
  // non-gridheist
  layoutType: 'fluid',
  layoutFluidValue: 100,
  layoutFixedValue: 1000
}
var config = _.extend({}, defaults);

// load any stored configs
if (localStorage && localStorage.getItem('config')) {
  var json = JSON.parse(localStorage.getItem('config'));
  if (json) {
    config = _.extend(config, json);
  }
}

// set the layout type
var updateLayoutType = function() {
  if (config.layoutType == 'fluid') {
    $('#gallery').css({marginLeft: config.layoutFluidValue, marginRight: config.layoutFluidValue, width: 'auto'});
  }
  else {
    $('#gallery').css({marginLeft: 'auto', marginRight: 'auto', width: config.layoutFixedValue});
  }
}

// dom => config
var updateConfig = function() {
  config.layoutType = $('select[name="layoutType"]').val();
  config.layoutFluidValue = parseInt( $('input[name="layoutFluidValue"]').val() )
  config.layoutFixedValue = parseInt( $('input[name="layoutFixedValue"]').val() )
  updateLayoutType();

  config.thumbBorder    = parseInt( $('input[name="thumbBorder"]').val() );
  config.thumbMinHeight = parseInt( $('input[name="thumbMinHeight"]').val() );
  config.thumbMaxHeight = parseInt( $('input[name="thumbMaxHeight"]').val() );
  $('#gallery').gridHeist('update', config);

  // update storage
  if (localStorage) {
    localStorage.setItem('config', JSON.stringify(config));
  }
}

// config => dom
var updateDom = function() {
  $('select[name="layoutType"]').val(config.layoutType);
  $('input[name="layoutFluidValue"]').val(config.layoutFluidValue)
    .filter('[data-slider]').simpleSlider('setValue', config.layoutFluidValue);
  $('input[name="layoutFixedValue"]').val(config.layoutFixedValue)
    .filter('[data-slider]').simpleSlider('setValue', config.layoutFixedValue);

  if (config.layoutType == 'fluid') {
    $('.layout-fluid').show() && $('.layout-fixed').hide();
  }
  else {
    $('.layout-fluid').hide() && $('.layout-fixed').show();
  }

  $('input[name="thumbBorder"]').val(config.thumbBorder)
    .filter('[data-slider]').simpleSlider('setValue', config.thumbBorder);
  $('input[name="thumbMinHeight"]').val(config.thumbMinHeight)
    .filter('[data-slider]').simpleSlider('setValue', config.thumbMinHeight);
  $('input[name="thumbMaxHeight"]').val(config.thumbMaxHeight)
    .filter('[data-slider]').simpleSlider('setValue', config.thumbMaxHeight);
}

// dom it!
$(function() {
  updateDom();
  updateLayoutType();

  // insert images
  $.getJSON('dynamic.json', function(data) {
    var $gallery = $('#gallery').html('');

    if (data && data.hits && data.hits.length) {
      for (var i=0; i<data.hits.length; i++) {
        var w = data.hits[i].webformatWidth,
            h = data.hits[i].webformatHeight,
            sm = data.hits[i].webformatURL,
            lg = data.hits[i].webformatURL;
        $gallery.append('<a class="thumb" href="'+lg+'"><img data-width="'+w+'" data-height="'+h+'" src="'+sm+'"/></a>');
      }

      $gallery.gridHeist(config);
    }
  });

  // sync sliders
  $('input[data-slider]').on('slider:changed', function(e, data) {
    $('[name="' + $(this).attr('name') + '"]').val(data.value);
  });

  // layout change
  $('input').on('change', updateConfig);
  $('select').on('change', function() {
    $('input').off('change', updateConfig);
    if ($(this).val() == 'fluid') {
      $('.layout-fluid').show() && $('.layout-fixed').hide();
      $('input[name="layoutFluidValue"]').val(100).filter('[data-slider]').simpleSlider('setValue', 100);
    }
    else {
      $('.layout-fluid').hide() && $('.layout-fixed').show();
      $('input[name="layoutFixedValue"]').val(1000).filter('[data-slider]').simpleSlider('setValue', 1000);
    }
    updateConfig();
    $('input').on('change', updateConfig);
  });

  // revert to defaults
  $('h2.reset').click(function() {
    alert('Reset to defaults');
    config = _.extend({}, defaults);
    if (localStorage) localStorage.clear();
    $('input').off('change', updateConfig);
    updateDom();
    $('input').on('change', updateConfig);
    $('#gallery').gridHeist('update', config);
  });

});


test/preloader.js
//
// preload image testing
//
var config = {
  thumbBorder:      10,
  thumbMinHeight:   100,
  thumbMaxHeight:   150,
  expander:         true,
  expandSideWidth:  200,
  preload:          true,
  preloadBefore:    1,
  preloadAfter:     4
}

// helper to get query params
var qs = function(key) {
  key = key.replace(/[*+?^$.\[\]{}()|\\\/]/g, "\\$&"); // escape RegEx meta chars
  var match = location.search.match(new RegExp("[?&]"+key+"=([^&]+)(&|$)"));
  return match && decodeURIComponent(match[1].replace(/\+/g, " "));
}

// dom it!
$(function() {

  // update form
  if (before = qs('preloadBefore') || config.preloadBefore) {
    $('input[name="preloadBefore"]').val(before);
    config.preloadBefore = parseInt(before);
  }
  if (after = qs('preloadAfter') || config.preloadAfter) {
    $('input[name="preloadAfter"]').val(after);
    config.preloadAfter = parseInt(after);
  }

  // insert images
  $.getJSON('preloader.json', function(data) {
    var $gallery = $('#gallery').html('');

    if (data && data.hits && data.hits.length) {
      var cb = Math.round(new Date().getTime() / 1000); //bust the cache
      for (var i=0; i<data.hits.length; i++) {
        var w = data.hits[i].webformatWidth,
            h = data.hits[i].webformatHeight,
            sm = data.hits[i].webformatURL + '?_cb=' + cb,
            lg = data.hits[i].bigformatURL + '?_cb=' + cb;
        $gallery.append('<a class="thumb" href="'+lg+'"><img data-width="'+w+'" data-height="'+h+'" src="'+sm+'"/></a>');
      }

      $gallery.gridHeist(config);
    }
  });

});


gridheist.js
// ==============================================================
// gridheist.js v0.0.3
// A jQuery plugin to hijack an image gallery, making it great.
// http://github.com/cavis/gridheist
// ==============================================================
// Copyright (c) 2013 Ryan Cavis
// Licensed under http://en.wikipedia.org/wiki/MIT_License
// ==============================================================
// Generated by CoffeeScript 1.6.3
(function() {
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __slice = [].slice;

  (function($, window) {
    var GridHeist;
    GridHeist = (function() {
      GridHeist.prototype["public"] = {
        update: true
      };

      GridHeist.prototype.defaults = {
        thumbSelector: '> *',
        thumbBorder: 10,
        thumbMinHeight: 200,
        thumbMaxHeight: null,
        preloadImages: true,
        expandHeight: 300,
        expandSideWidth: 200,
        expandSideRender: false
      };

      function GridHeist(el, options) {
        this.keyHandler = __bind(this.keyHandler, this);
        this.moveRight = __bind(this.moveRight, this);
        this.moveLeft = __bind(this.moveLeft, this);
        this.clickHandler = __bind(this.clickHandler, this);
        var _this = this;
        this.options = $.extend({}, this.defaults, options);
        if (this.options.thumbMaxHeight && this.options.thumbMaxHeight < this.options.thumbMinHeight) {
          this.options.thumbMinHeight = this.options.thumbMaxHeight;
        }
        this.$el = $(el);
        this.$el.addClass('gridheist-gallery');
        this.$el.on('click', '.gridheist-thumb', this.clickHandler);
        this.$el.on('click', '.gridheist-left', function() {
          return _this.moveLeft();
        });
        this.$el.on('click', '.gridheist-right', function() {
          return _this.moveRight();
        });
        this.$el.on('click', '.gridheist-close', function() {
          return _this.closeExpander();
        });
        this.update();
        $(window).resize(function() {
          if (_this.width !== _this.$el.width()) {
            return _this.doLayout();
          }
        });
      }

      GridHeist.prototype.update = function() {
        this.$thumbs = this.$el.find(this.options.thumbSelector);
        this.$thumbs.addClass('gridheist-thumb');
        return this.doLayout();
      };

      GridHeist.prototype.doLayout = function() {
        var margin;
        this.closeExpander();
        this.width = this.$el.width();
        this.rows = [];
        margin = "0 " + this.options.thumbBorder + "px " + this.options.thumbBorder + "px 0";
        this.$thumbs.css('margin', margin);
        return this.processThumb(0, 0, 0);
      };

      GridHeist.prototype.processThumb = function(idx, rowIdx, rowWidth) {
        var $img, $thumb,
          _this = this;
        if (idx < this.$thumbs.length) {
          $thumb = this.$thumbs.eq(idx);
          $img = $thumb.find('img');
          return this.getDimensions($img, function(imgWidth, imgHeight) {
            if (!_this.rows[rowIdx]) {
              _this.rows[rowIdx] = {
                thumbs: [],
                images: [],
                widths: [],
                heights: []
              };
            }
            if (_this.options.thumbMaxHeight && imgHeight > _this.options.thumbMaxHeight) {
              imgWidth = (_this.options.thumbMaxHeight / imgHeight) * imgWidth;
              imgHeight = _this.options.thumbMaxHeight;
              $img.attr('width', imgWidth);
              $img.attr('height', imgHeight);
            }
            _this.rows[rowIdx].thumbs.push($thumb);
            _this.rows[rowIdx].images.push($img);
            _this.rows[rowIdx].widths.push(imgWidth);
            _this.rows[rowIdx].heights.push(imgHeight);
            $thumb.data('row', rowIdx);
            if (rowWidth !== 0) {
              rowWidth += _this.options.thumbBorder;
            }
            rowWidth += imgWidth;
            if (rowWidth >= _this.width) {
              _this.layoutRow(rowIdx, rowWidth);
              return _this.processThumb(idx + 1, rowIdx + 1, 0);
            } else {
              return _this.processThumb(idx + 1, rowIdx, rowWidth);
            }
          });
        } else {
          if (this.rows[rowIdx]) {
            this.layoutRow(rowIdx, rowWidth);
          }
          if (this.options.preloadImages) {
            return this.preload();
          }
        }
      };

      GridHeist.prototype.layoutRow = function(rowIdx, rowWidth) {
        var $thumb, center, h, idx, margin, minHeight, rowDebt, thumbDebt, vertic, _i, _j, _len, _len1, _ref, _ref1, _results;
        minHeight = 9999;
        _ref = this.rows[rowIdx].heights;
        for (idx = _i = 0, _len = _ref.length; _i < _len; idx = ++_i) {
          h = _ref[idx];
          if (h < minHeight) {
            minHeight = h;
          }
        }
        minHeight = Math.max(this.options.thumbMinHeight, minHeight);
        rowDebt = Math.max(rowWidth - this.width, 0);
        _ref1 = this.rows[rowIdx].thumbs;
        _results = [];
        for (idx = _j = 0, _len1 = _ref1.length; _j < _len1; idx = ++_j) {
          $thumb = _ref1[idx];
          thumbDebt = Math.ceil(rowDebt / (this.rows[rowIdx].thumbs.length - idx));
          rowDebt -= thumbDebt;
          $thumb.css('width', this.rows[rowIdx].widths[idx] - thumbDebt);
          $thumb.css('height', minHeight);
          center = Math.floor(thumbDebt / 2);
          vertic = Math.floor((this.rows[rowIdx].heights[idx] - minHeight) / 2);
          this.rows[rowIdx].images[idx].css('margin-left', -center);
          this.rows[rowIdx].images[idx].css('margin-top', -vertic);
          if (idx === this.rows[rowIdx].thumbs.length - 1) {
            margin = "0 0 " + this.options.thumbBorder + "px 0";
            _results.push($thumb.css('margin', margin));
          } else {
            _results.push(void 0);
          }
        }
        return _results;
      };

      GridHeist.prototype.getDimensions = function($img, callback) {
        var h, tmp, w;
        if ((w = $img.data('width')) && (h = $img.data('height'))) {
          return callback(w, h);
        } else {
          tmp = new Image();
          return $(tmp).load(function() {
            $img.data('width', tmp.width);
            $img.data('height', tmp.height);
            return callback(tmp.width, tmp.height);
          }).error(function() {
            console.error('An error occurred and your image could not be loaded.  Please try again.');
            return callback(100, 100);
          }).attr({
            src: $img.attr('src')
          });
        }
      };

      GridHeist.prototype.preload = function() {
        return this.$thumbs.each(function(idx, thumb) {
          return (new Image()).src = $(thumb).attr('href');
        });
      };

      GridHeist.prototype.clickHandler = function(e) {
        var $thumb;
        e.preventDefault();
        $thumb = $(e.currentTarget);
        if (!$thumb.hasClass('expanded')) {
          return this.expandThumb($thumb);
        }
      };

      GridHeist.prototype.moveLeft = function(e) {
        var $curr, $prev;
        $curr = this.$el.find('.gridheist-thumb.expanded');
        $prev = $curr.prevAll('.gridheist-thumb').first();
        if ($prev.length > 0) {
          return this.expandThumb($prev, 'up');
        } else {
          return this.closeExpander();
        }
      };

      GridHeist.prototype.moveRight = function(e) {
        var $curr, $next;
        $curr = this.$el.find('.gridheist-thumb.expanded');
        $next = $curr.nextAll('.gridheist-thumb').first();
        if ($next.length > 0) {
          return this.expandThumb($next, 'down');
        } else {
          return this.closeExpander();
        }
      };

      GridHeist.prototype.keyHandler = function(e) {
        var code;
        code = e.keyCode || e.which;
        if (code === 37) {
          this.moveLeft();
        }
        if (code === 39) {
          this.moveRight();
        }
        if (code === 27) {
          return this.closeExpander;
        }
      };

      GridHeist.prototype.expandThumb = function($thumb, doScroll) {
        var $lastThumb, align, bordr, diff, imgSrc, rowIdx, scrollTo, width;
        if (doScroll == null) {
          doScroll = null;
        }
        rowIdx = $thumb.data('row');
        imgSrc = $thumb.attr('href');
        $lastThumb = this.rows[rowIdx].thumbs[this.rows[rowIdx].thumbs.length - 1];
        scrollTo = null;
        this.$el.find('.gridheist-thumb.expanded').removeClass('expanded');
        $thumb.addClass('expanded');
        if (doScroll) {
          scrollTo = $('body').scrollTop();
          diff = this.options.thumbBorder + $thumb.height();
          if (doScroll === 'up') {
            scrollTo -= diff;
          } else {
            scrollTo += diff;
          }
        }
        if (!this.$expander) {
          doScroll = false;
          this.$expander = $('<div class="gridheist-expander"></div>');
          $(document).on('keydown', this.keyHandler);
        } else if (!this.$expander.prev().is($lastThumb)) {
          this.$expander.remove();
          this.$expander = $('<div class="gridheist-expander"></div>');
        } else {
          doScroll = false;
          this.$expander.html('');
        }
        this.$expander.css('margin-bottom', this.options.thumbBorder);
        this.$expander.css('height', this.options.expandHeight);
        if (this.options.expandSideRender) {
          width = this.options.expandSideWidth;
          this.$expander.append("<div class=\"gridheist-expander-side\" style=\"width:" + width + "px\">\n  <div class=\"gridheist-expander-seperator\"></div>\n  <div class=\"gridheist-expander-content\">\n    " + (this.options.expandSideRender($thumb)) + "\n  </div>\n</div>");
        }
        align = $thumb.position().left + ($thumb.width() / 2);
        bordr = this.options.thumbBorder;
        this.$expander.append("<div class=\"gridheist-expander-arrow\" style=\"left:" + align + "px;\n  border-width:" + bordr + "px; margin-left:-" + bordr + "px;\">\n</div>\n<div class=\"gridheist-left\">&lsaquo;</div>\n<div class=\"gridheist-right\">&rsaquo;</div>\n<div class=\"gridheist-close\">&times;</div>");
        this.$expander.find('.gridheist-left, .gridheist-right, .gridheist-close').attr('unselectable', 'on').css('user-select', 'none').on('selectstart', false);
        this.$expander.append("<div class=\"gridheist-expander-img\">\n  <img src=\"" + imgSrc + "\"/>\n</div>");
        if (!this.$expander.parent().length) {
          $lastThumb.after(this.$expander);
        }
        if (doScroll) {
          return $('body').scrollTop(scrollTo);
        }
      };

      GridHeist.prototype.closeExpander = function() {
        this.$el.find('.expanded').removeClass('expanded');
        if (this.$expander) {
          this.$expander.remove();
          this.$expander = void 0;
          return $(document).off('keydown', this.keyHandler);
        }
      };

      return GridHeist;

    })();
    return $.fn.extend({
      gridHeist: function() {
        var args, option;
        option = arguments[0], args = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
        return this.each(function() {
          var $this, data;
          $this = $(this);
          data = $this.data('gridHeist');
          if (!data) {
            $this.data('gridHeist', (data = new GridHeist(this, option)));
          }
          if (typeof option === 'string' && data["public"][option]) {
            return data[option].apply(data, args);
          }
        });
      }
    });
  })(window.jQuery, window);

}).call(this);
